### PR TITLE
[BUGFIX] Fix `willDestroy` on class helpers

### LINKED ISSUE #18830: Class-based helpers' willDestroy() hook called multiple times in tests
In Ember 3.17, class-based helpers' `willDestroy()` hooks are being called twice from acceptance tests.

See [this twiddle](https://ember-twiddle.com/212dc178f1da9dc03ad270fbec3ea60f?openFiles=helpers.my-helper%5C.js%2C) for a repro as well as stack traces to the two calls.

This does not happen if you change the Ember.js version to 3.16, so it's new in 3.17. This is causing problems in my helpers that do things like manage the lifetime of `@ember/object/events` listeners (it double-removes and the second removal throws the listener-not-registered assertion).

### PR DESCRIPTION
**Pull request by pzuraq**
Class based helpers were having `willDestroy` be called twice, once by
the Glimmer VM's built-in destroyables implementation, and once
scheduled by `destroy` (which is the correct destruction). This is
because we were registering the class itself directly as the
destroyable, and Glimmer VM detected the `willDestroy` function on it
and called it eagerly.

This fix creates an intermediate destructor object with a `destroy()`
hook, but long term we need to finish the refactors to destroyables in
the VM and make it so these types of collisions can't happen again.

Added a basic lifecycle test for helpers to prevent this specific issue
from reoccuring.

Fixes #18830

### CODE REVIEW

**Comment by rwjblue** [comment]
Can you update this to `[BUGFIX release]`?

File: yarn.lock [hunk 1]
```diff
@@ -2,7 +2,7 @@
 # yarn lockfile v1
 
 
-"@babel/code-frame@^7.0.0", "@babel/code-frame@^7.5.5", "@babel/code-frame@^7.8.3":
+"@babel/code-frame@^7.0.0", "@babel/code-frame@^7.8.3":
```
Comment (rwjblue) [inline yarn.lock]: Can you reset this files changes? Seems unrelated to the other changes here...

Comment (pzuraq) [reply yarn.lock]: FWIW I think this just happened when I ran `yarn` locally`, unsure. But yes, will reset

**Review by rwjblue — APPROVED** [review]
